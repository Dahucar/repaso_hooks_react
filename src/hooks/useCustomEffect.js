import { useEffect, useRef, useState } from "react"

// https://www.breakingbadapi.com/api/episodes/60
export const useCustomEffect = ( url ) => {

    // el useRef mantendra el valor que asignado
    const isActive = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null })

    //efecto para determinar cuando este componente deja de estar activo
    useEffect(() => {
        return () => {
            isActive.current = false;
        }
    }, [  ])

    useEffect(() => {
        // logica a ejecutar
        setState({ data: null, loading: true, error: null });
        // cambios para probar hook useRef
        fetch( url )
        .then( resp => resp.json() )
        .then( data => {
            if(isActive.current){
                setState({
                    loading: false,
                    error: null,
                    data
                });
            }else{
                console.log('Componente inactivo -> error prevenido');
            }
        })
        .catch( () => {
            setState({
                data: null,
                loading: false,
                error: 'Error en la petici√≥n.',
            });
        });
    }, [ url ]);

    return state;
}
