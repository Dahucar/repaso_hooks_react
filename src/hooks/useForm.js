import { useState } from 'react'

export const useForm = ( initialState = {} ) => {
    
    const [values, setValues] = useState(initialState);

    const handleInputChangue = ({ target }) => {
        setValues({
            ...values,
            [ target.name ]: target.value
        });
    } 

    return [ values, handleInputChangue ];

}
