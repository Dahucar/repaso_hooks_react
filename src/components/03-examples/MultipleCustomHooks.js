import React from 'react'
import { useCustomEffect } from '../../hooks/useCustomEffect'
import { useCustomCounter } from '../01-useState/useCustomCounter';

export const MultipleCustomHooks = () => {

    const { state:counter, increment, decrement } = useCustomCounter(1);
    const { loading, data } = useCustomEffect( 'https://www.breakingbadapi.com/api/quotes/'+counter );

    /*
        Resultado de la API.
        "quote": "I am not in danger, Skyler. I am the danger!",
        "author": "Walter White",
        "series": "Breaking Bad"
    */
    // data por defecto tendra valor "null" al agregarle "!data" será true y con "!!data" será false. 
    // por ende si es false obtenemos el resultado del array
    const { quote, author, series } = !!data && data[0];

    return (
        <div className="container card">
            <div className="card-header">
                <h2>Custom Hooks with BreakingBad quotes</h2>
            </div>
            <div className="card-body">
                {
                    loading ? 
                    (
                        <div className="spinner-grow text-success" role="status">
                            <span className="sr-only">Loading...</span>
                        </div>
                    ) : 
                    (
                        <div>
                            <blockquote className="blockquote text-right">
                                <p className="mb-0"> quote - { counter } </p>
                                <p className="mb-0"> { quote } </p>
                                <footer className="blockquote-footer">{ author } <cite title="Source Title">{ series }</cite></footer>
                            </blockquote>
                            <button className="btn btn-success"
                                onClick={ increment }>
                                Siguiente
                            </button>
                            <button className="btn btn-danger"
                                onClick={ decrement }>
                                Volver
                            </button>
                        </div> 
                    )
                }
                
            </div>
        </div>
    )
}
