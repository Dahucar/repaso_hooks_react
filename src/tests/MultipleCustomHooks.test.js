import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../components/03-examples/MultipleCustomHooks';
import { useCustomEffect } from '../hooks/useCustomEffect';
//importaciones usadas en <MultipleCustomHooks />
// jest.mock: simula el uso del elemento dado.
jest.mock('../hooks/useCustomEffect');

describe('Pruebas en <MultipleCustomHooks />', () => {
    test('Debe hacer match con la instantanea.', () => {
        useCustomEffect.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        const wrapper = shallow( <MultipleCustomHooks /> );
        expect( wrapper ).toMatchSnapshot();
    });
    test('Demo de mostrar la informacion obtenida', () => {
        useCustomEffect.mockReturnValue({
            data: [{
                author: 'Daniel',
                quote: 'Una frase random.'
            }],
            loading: false,
            erros: null
        });    
        const wrapper = shallow( <MultipleCustomHooks /> );  
        expect( wrapper.find('.spinner-grow').exists() ).toBe( false );
        expect( wrapper.find('.blockquote-footer').text().trim() ).toBe( 'Daniel' );
    });
});
