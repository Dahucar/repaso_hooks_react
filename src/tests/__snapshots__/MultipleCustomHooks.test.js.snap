// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe hacer match con la instantanea. 1`] = `
<div
  className="container card"
>
  <div
    className="card-header"
  >
    <h2>
      Custom Hooks with BreakingBad quotes
    </h2>
  </div>
  <div
    className="card-body"
  >
    <div
      className="spinner-grow text-success"
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
